src/katgpucbf/xbgpu/engine.py

```diff
@@ -117,6 +127,417 @@ def reset(self, timestamp: int = 0) -> None:
+        self._rx_item_queue: asyncio.Queue[RxQueueItem | None] = engine.monitor.make_queue(
+            f"{name}.rx_item_queue", self.n_rx_items
+        )
+        self._tx_item_queue: asyncio.Queue[TxQueueItem | None] = engine.monitor.make_queue(
+            f"{name}.tx_item_queue", self.n_tx_items
+        )
+        self._tx_free_item_queue: asyncio.Queue[TxQueueItem] = engine.monitor.make_queue(
+            f"{name}.tx_free_item_queue", self.n_tx_items
+        )
+
+    def add_rx_item(self, item: RxQueueItem) -> None:
+        """Append a newly-received :class:`RxQueueItem` to the :attr:`_rx_item_queue`."""
+        self._rx_item_queue.put_nowait(item)
+
+    def shutdown(self) -> None:
+        """Start a graceful shutdown after the final call to :meth:`add_rx_item`."""
+        self._rx_item_queue.put_nowait(None)
+
+    @abstractmethod
+    async def gpu_proc_loop(self) -> None:
+        """Perform all GPU processing of received data in a continuous loop.
+
+        This function should do the following:
+        - Get an RxQueueItem off the rx_item_queue
+        - Ensure it is not a NoneType value (indicating shutdown sequence)
+        - await any outstanding events associated with the RxQueueItem
+        - Apply GPU processing to data in the RxQueueItem
+
+            - Bind input buffer(s) accordingly
+
+        - Obtain a free TxQueueItem from the tx_free_item_queue
+
+            - Add event marker to wait for the proc_command_queue
+            - Put the prepared TxQueueItem on the tx_item_queue
+
+        NOTE: An initial TxQueueItem needs to be obtained from the tx_free_item_queue
+        for the first round of processing:
+
+        - The gpu_proc_loop requires logic to decipher the timestamp of the
+          first heap output.
+        - It also provides an opportunity to bind buffers before processing is queued.
+        """
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    async def sender_loop(self) -> None:
+        """Send heaps to the network in a continuous loop.
+
+        This function should do the following:
+        - Get a TxQueueItem from the tx_item_queue
+        - Ensure it is not a NoneType value (indicating shutdown sequence)
+        - Wait for events on the item to complete (likely GPU processing)
+        - Wait for an available heap buffer from the send_stream
+        - Asynchronously Transfer/Download GPU buffer data into the heap buffer in system RAM
+
+            - Wait for the transfer to complete, before
+
+        - Transmit heap buffer onto the network
+        - Place the TxQueueItem back on the tx_free_item_queue once complete
+        """
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def capture_enable(self, enable: bool = True) -> None:
+        """Enable/Disable the transmission of this data product's stream."""
+        raise NotImplementedError
```

Comment:
Since you've used `#pragma: nocover` a few lines further up, you probably want it here too.